src/libopensc/card-piv.c

```diff
@@ -2440,15 +2440,16 @@ static int piv_select_file(sc_card_t *card, const sc_path_t *in_path,
 	pathlen = in_path->len;
 
 	/* only support single EF in current application */
+	/* 
+	 * PIV emulates files, and only does so becauses sc_pkcs15_* uses
+	 * select_file and read_binary. The emulation adds path emulated structures
+	 * so piv_select_file will find it.
+	 * there is no dir. Only direct access to emulated files
+	 * thus opensc-tool and opensc-explorer can not read the emulated files
+	 */
 
 	if (memcmp(path, "\x3F\x00", 2) == 0) {
-		if (pathlen == 2)   {
-			r = piv_select_aid(card, piv_aids[0].value, piv_aids[0].len_short, NULL, NULL);
-			SC_TEST_RET(card->ctx, SC_LOG_DEBUG_NORMAL, r, "Cannot select PIV AID");
-
-			SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_NORMAL, r);
-		}
-		else if (pathlen > 2) {
+		if (pathlen > 2) {
```

Comment:
The problem was the SC_TEST_RET would return 0 without filling in *file which caused the segfault later in
opensc-tool. In the comments above the removed code:

         /* only support single EF in current application */
         /*
          * PIV emulates files, and only does so because sc_pkcs15_* uses
          * select_file and read_binary. The emulation adds path emulated structures
          * so piv_select_file will find it.
          * there is no dir. Only direct access to emulated files
          * thus opensc-tool and opensc-explorer can not read the emulated files
          */

Only selected paths are supported as defined in the code. piv_find_obj_by_containerid called below the code does this.

The NIST 800-73-3 part 2 defines a subset of ISO 7816-4 commands to be used. It makes no assumption
that any other commands need to be supported. It has "SELECT" command which happens to be 00 A4 04 00 Lc AID Le
All reading of data from the card is done using "GET DATA"

The PIV driver was written to stick to the standards as close as possible and not assume anything else
about the card and present a PKCS#11 interface. (Because of the way OpenSC uses PKCS#15 some of the sc_pkcs15
routines also work but not all.)



On 10/16/2015 4:44 AM, Frank Morgner wrote:
> In src/libopensc/card-piv.c <https://github.com/OpenSC/OpenSC/pull/579#discussion_r42225468>:
>
>>
>>  	if (memcmp(path, "\x3F\x00", 2) == 0) {
>> -		if (pathlen == 2)   {
>> -			r = piv_select_aid(card, piv_aids[0].value, piv_aids[0].len_short, NULL, NULL);
>> -			SC_TEST_RET(card->ctx, SC_LOG_DEBUG_NORMAL, r, "Cannot select PIV AID");
>> -
>> -			SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_NORMAL, r);
>> -		}
>> -		else if (pathlen > 2) {
>> +		if (pathlen > 2) {
>
> The code you remove /should/ work. What was the problem?
>
> The change you introduced will /always/ remove the first FID. This will not work with a path relative to the current DF.
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/pull/579/files#r42225468>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>